### ISSUE TITLE
RL rl_6c — l'échantillonnage d'importance jamais nommé : le ratio implémenté, tracé et mis en exercice, la réutilisation par epochs étiquetée « On-policy »

### ISSUE THREAD

**Issue by myia-ai-01** [jsboige/CoursIA#12517]
## Constat

Mesure sur les **21 notebooks** de `MyIA.AI.Notebooks/RL/` (code et markdown) :

```
importance sampling              code=0  md=0
échantillonnage d'importance     code=0  md=0
poids d'importance               code=0  md=0
importance weight                code=0  md=0
correction off-policy            code=0  md=0
behavior policy / target policy  code=0  md=0
ESS / effective sample size      code=0  md=0
```

Contrôles positifs sur la même mesure — le vocabulaire environnant est massivement présent, la requête n'est donc pas hors sujet :

```
ratio        code=100  md=211  nb=21/21
off-policy             md=27   nb=8
on-policy              md=9    nb=4
clip         code=50   md=49   nb=8
surrogate    code=9    md=13
KL           code=17   md=55   nb=7
old_log      code=9            (rl_6c, rl_6e)
```

Le ratio est écrit, implémenté, tracé et mis en exercice — dans trois notebooks :

- `rl_6c_ppo_from_scratch.ipynb` c4 pose la formule : « L'idee centrale de PPO est le **ratio de probabilite** : $r_t(\theta) = \frac{\pi_\theta(a_t | s_t)}{\pi_{\theta_{old}}(a_t | s_t)}$ »
- c10 le calcule : `ratio = torch.exp(new_log_probs - mb_old_log_probs)`
- c6 le trace sur `np.linspace(0.5, 1.5, 200)` contre les objectifs clippé et non clippé
- c19/c20 en font l'exercice 1 : `clipped_surrogate_loss(ratio, advantages, epsilon=0.2)`, docstring `ratio: r_t(theta) = pi_nouveau / pi_ancien`
- `rl_6e_grpo_from_scratch.ipynb` c7 le redonne : « **Ratio clippe (PPO)** : $\frac{\pi_\theta(a|s)}{\pi_{\theta_{old}}(a|s)}$ »
- `rlpt_1_ppo_lm_rlhf.ipynb` c11 l'écrit encore : « déviation au ratio $\rho_t = \pi_\theta / \pi_{\theta_{\text{old}}}$ »

Jamais l'objet n'est nommé. La cellule 4 le présente comme **une idée de conception** (« l'idée centrale de PPO »), pas comme le terme de correction qu'il est.

### La ligne où le concept manquant aurait changé le texte

`rl_6c` c23, tableau A2C vs PPO :

| Aspect | A2C | PPO |
|--------|-----|-----|
| Mises a jour | 1 par trajectoire | Plusieurs epochs par trajectoire |
| Echantillonnage | **On-policy, 1 utilisation** | **On-policy, re-utilisation via epochs** |

et c27, point clé n°2 : « Les mini-lots et epochs permettent de **reutiliser les données collectees**, ameliorant l'efficacite d'echantillonnage par rapport a A2C ».

L'implémentation fait bien ce que le tableau annonce — `ppo_epochs=4` au constructeur (c10) et `for _ in range(self.ppo_epochs):` dans `update()`. Mais dès la deuxième epoch, $\pi_\theta \neq \pi_{\theta_{old}}$ : les gradients suivants sont calculés sur des données que la politique courante n'a **pas** produites. Ce n'est plus on-policy au sens strict, et c'est exactement pourquoi le ratio est là.

Le notebook énonce le bénéfice (efficacité d'échantillonnage par réutilisation), énonce le mécanisme (le ratio, clippé), et n'énonce jamais le lien — au point d'étiqueter « On-policy » la case même où la correction est requise.

### Le contrôle interne au dépôt : une série parallèle le nomme, sur le même algorithme

Ce n'est pas un arbitrage de niveau. Mesure repo-wide sur les **1089 notebooks** : `importance sampling` apparaît dans **3** notebooks, dont deux enseignent le même sujet dans une autre série —

- `GenAI/PostTraining/PT_08_grpo_from_scratch_toy_env.ipynb` c9 : « PPO (Schulman et al., 2017) introduit le **ratio** d'importance sampling » ; c24 : « d'importance sampling, les epochs internes — tout est identique à PPO. »
- `GenAI/PostTraining/PT_10_gae_from_scratch_toy_env.ipynb` c4 : « Pour PPO, on aurait 4 value heads + clipped objective + importance sampling. **Ici on reste minimal** : 1 value head, pas de clipping, pas de ratio. »
- `Probas/PyMC/PyMC-10-Model-Selection.ipynb` c4/c27 : PSIS-LOO, référencé à Vehtari, Gelman & Gabry (2017), avec le diagnostic de Pareto k.

`PT_08` est le GRPO-from-scratch de GenAI/PostTraining ; `rl_6e_grpo_from_scratch` est celui de RL. Le premier nomme le concept, le second non. `PT_10` le nomme même pour dire qu'il l'exclut délibérément. Le dépôt sait donc écrire ce paragraphe : il manque précisément dans la série qui en a le plus besoin.

### Ce que la série pose sans le concept qui l'encadre

`rl_9_offline_rl.ipynb` — l'offline RL est le cadre où la correction d'importance et sa variance sont *le* problème central :

```
distribution shift           md=1
politique de comportement    code=1  md=6
pessimis                     md=2
importance sampling / ESS    0
```

## Pourquoi ça compte

- **La réutilisation par epochs n'est jamais justifiée.** Le gradient de politique est une espérance sous $\pi_\theta$. `update()` la calcule sur des trajectoires tirées sous $\pi_{\theta_{old}}$, quatre fois de suite. L'identité qui autorise cela tient en une ligne : $\mathbb{E}_{\pi_\theta}[f] = \mathbb{E}_{\pi_{old}}\!\left[\frac{\pi_\theta}{\pi_{old}} f\right]$. Sans elle, l'étudiant qui vient de comprendre REINFORCE (`rl_6_policy_gradient`) ne peut pas savoir pourquoi PPO a le droit de faire 4 epochs quand A2C n'en fait qu'une — et le tableau c23 lui affirme que les deux sont « on-policy ».
- **Le clip perd la moitié de sa raison d'être.** c4 le justifie par « une seule mise a jour peut changer radicalement la politique » : vrai, mais incomplet. Le clip est aussi, et d'abord, un **contrôle de variance sur l'estimateur d'importance** — quand $\pi_\theta$ s'éloigne de $\pi_{old}$, le ratio explose et l'estimateur devient inutilisable bien avant que la politique ne soit « radicalement » changée. C'est ce qui explique à la fois l'ordre de grandeur $\epsilon \approx 0.2$ et le fait que PPO re-collecte après quelques epochs seulement.
- **Trois notebooks avals reposent dessus.** `rl_6e` (GRPO) réutilise le même ratio ; `rlpt_1` (PPO-LM/RLHF) l'écrit en $\rho_t$ juste à côté d'une pénalité KL — or la KL vers une politique de référence et le poids d'importance vers $\pi_{old}$ sont deux objets distincts, que rien dans la série ne distingue pour l'étudiant. `rl_9_offline_rl` pose le *distribution shift* et la *politique de comportement* sans le poids qui les relie.
- **Référence canonique.** Sutton & Barto, *Reinforcement Learning* (2e éd.), §5.5 « Off-policy Prediction via Importance Sampling » et §7.3 : l'échantillonnage d'importance y est le mécanisme de base de tout l'off-policy, introduit avant DQN comme avant les méthodes de gradient. Schulman et al. (2017) §3 dérive $L^{CLIP}$ depuis l'objectif *surrogate* de TRPO, lui-même construit sur ce ratio. Une série qui va de CartPole à GRPO, SAC, offline RL et DPO n'a pas écarté le concept par choix de niveau : il est plus élémentaire que ce qu'elle couvre déjà.

## Correction attendue

Cible : `rl_6c_ppo_from_scratch.ipynb`. Un notebook, une PR.

1. Ajouter, **avant** la §2 actuelle (« L'objectif surrogate clippe »), une section « Pourquoi un ratio ? L'échantillonnage d'importance » : poser $\mathbb{E}_{x\sim p}[f(x)] = \mathbb{E}_{x\sim q}\!\left[\frac{p(x)}{q(x)}f(x)\right]$, puis l'instancier sur $p = \pi_\theta$ et $q = \pi_{\theta_{old}}$ de façon à faire **tomber** $r_t$ au lieu de le poser.
2. Relier explicitement à `ppo_epochs` : la 1ʳᵉ epoch est on-policy ($r_t = 1$ partout), les suivantes ne le sont plus — c'est la réutilisation que le ratio paie.
3. **Une mesure, pas une affirmation.** Ajouter une cellule exécutée qui, sur un batch réellement collecté par l'agent, calcule la distribution des ratios **à chaque epoch** — epoch 1 : tous à 1.0 ; epochs 2 à 4 : dispersion croissante — et l'affiche. Sortie committée. C'est la démonstration que le concept est opérant dans leur propre code, pas une note théorique posée à côté.
4. Ajouter une mesure de dégradation de l'estimateur : tracer l'ESS $\left(\sum_i w_i\right)^2 / \sum_i w_i^2$ des ratios par epoch et commenter sa chute. (Variante plus lourde si le volume le permet : faire diverger volontairement les politiques par quelques pas à `lr` élevé et montrer l'explosion de la variance du gradient estimé, puis son bornage par le clip.)
5. Corriger la ligne « Echantillonnage » du tableau c23 : PPO n'est pas « On-policy, re-utilisation via epochs ». Formulation juste : « on-policy à la collecte, corrigé par ratio d'importance sur les epochs suivantes ».
6. Compléter le point clé n°2 de c27 par la raison de la validité de la réutilisation — une phrase suffit.
7. Ajouter un exercice (stub C.1, sans `raise NotImplementedError`) : par exemple implémenter `effective_sample_size(ratios)`, ou calculer à la main le ratio pour deux distributions catégorielles données et dire ce qu'il advient de l'estimateur lorsqu'une action rare sous $\pi_{old}$ devient probable sous $\pi_\theta$.

## Critères d'acceptation

- [ ] « échantillonnage d'importance » (ou « importance sampling ») apparaît en markdown de `rl_6c`, avec l'identité $\mathbb{E}_p[f] = \mathbb{E}_q[(p/q)\,f]$ écrite — aujourd'hui **0 occurrence sur les 21 notebooks de la série**.
- [ ] Le lien ratio ↔ `ppo_epochs` est écrit : pourquoi la réutilisation des données est licite.
- [ ] **Une cellule exécutée** montre la distribution (ou l'ESS) des ratios par epoch, avec sortie committée : epoch 1 à 1.0, dispersion ensuite. Une section théorique seule ne satisfait pas ce critère.
- [ ] La ligne « Echantillonnage » du tableau c23 ne qualifie plus la réutilisation par epochs de « On-policy » sans mention de la correction.
- [ ] Le clip est justifié **aussi** comme contrôle de variance de l'estimateur d'importance, pas seulement comme trust region.
- [ ] Un exercice ajouté, stub conforme C.1 — le stub existant c20 (`return None  # TODO etudiant`) est conforme et le reste.
- [ ] Ré-exécution complète C.2 : `execution_count != null` et outputs cohérents sur les 10 cellules code (état actuel : **10/10 exécutées, 0 `exec_null`** — cet état ne doit pas régresser).
- [ ] Aucune suppression de contenu : c4 (formule + clip), c6 (visualisation), c10 (agent PPO), c19/c20 (exercice), c23 et c27 restent en place.

## Complémentarité

Un notebook, une PR. Cible unique `MyIA.AI.Notebooks/RL/rl_6c_ppo_from_scratch.ipynb` — c'est le premier endroit de la série où le ratio apparaît, donc le seul où le concept doit être introduit.

**Hors périmètre, à traiter séparément si utile** : `rl_6e_grpo_from_scratch.ipynb` (c7) et `rlpt_1_ppo_lm_rlhf.ipynb` (c11) réécrivent le ratio ; une fois `rl_6c` corrigé, un simple renvoi y suffira probablement — à évaluer alors, pas à empiler ici. `rl_9_offline_rl.ipynb` (distribution shift et politique de comportement posés sans poids d'importance) est un sujet distinct et plus lourd : il mérite sa propre issue, pas une section dans celle-ci.

**Ne remet en cause aucun contenu existant.** L'implémentation PPO de c10 est correcte (GAE, normalisation des avantages, mini-lots, clip du gradient, `min` des deux surrogates) ; la visualisation c6 est juste ; l'exercice c19/c20 est bien posé. L'ajout est une section amont plus la correction de deux formulations, pas une réécriture.

**Distincte de** #12431 (`rl_1_intro_cartpole`, courbe d'apprentissage manquante) et #12436 (`rl_5`, atome Q non isolé) : aucune des deux ne touche `rl_6c` ni la famille PPO. Aucune PR ouverte ne touche ce notebook — vérifié : #12495 → `rl_1_intro_cartpole`, #12487 et #12481 → `rl_5_mdp_dp_qlearning`.

**Comment by jsboigeEpita** [comment]
[CLAIMED] #12517 — myia-po-2023:CoursIA 2026-08-23 — rl_6c : section échantillonnage d'importance (identité E_p[f]=E_q[(p/q)f]), mesure des ratios par epoch + ESS sur batch réellement collecté, tableau c23 corrigé, clip justifié comme contrôle de variance, exercice ESS stub C.1 (Grain MED/notebook-python, See #12517)

**Comment by jsboige** [comment]
[DELIVERED] lane myia-po-2023:CoursIA -- PR #12521 (feature/12517-rl6c-importance-sampling, commits 988c71be23 + 0aa3252dc9) — marquage de la livraison du 2026-08-23 (le commentaire [DELIVERED] manquait ; le claim ci-dessus reste le mien, livré par cette PR).

Substance : rl_6c nomme l'échantillonnage d'importance — identité E_p[f]=E_q[(p/q)f] dérivée (r_t tombe au lieu d'être posé), mesure des ratios par epoch sur batch réellement collecté (epoch 1 à 1.0, dispersion ensuite), ESS par epoch, tableau A2C/PPO corrigé (« on-policy à la collecte, corrigé par ratio d'importance sur les epochs suivantes »), clip justifié aussi comme contrôle de variance, exercice ESS stub C.1. Re-exec complet C.2 (10/10 code cells, 0 exec null).

Note pour le reviewer : le check lane-collision de #12521 est actuellement rouge à cause du bug #12719 (le claim fondateur ci-dessus porte une date nue → lane fantôme `myia-po-2023:CoursIA 2026-08-23` ≠ lane déclarée → auto-blocage). Fix du parser = PR #12725 (en CI). Après merge de #12725 + update-branch de #12521, le garde passe. Fallback : le claim passe stale (>48h) le 2026-08-25T03:53Z et cesse de bloquer.

**Comment by jsboige** [comment]
[RELEASED] lane myia-po-2023:CoursIA 2026-08-23 -- cle exacte du claim mal forme (date nue, bug #12719). Livre par PR #12521.
